### PR TITLE
Qubits shouldn't be mutable

### PR DESCRIPTION
**Pull request by bt3gl**
Resolves # #2356 

Uses `dataclasses` with `frozen=True` to make LineQubit, GridQubit and NamedQubit immutable.

### NamedQubit

```
import cirq
q1 = cirq.NamedQubit('awesome_qubit')
circuit = cirq.Circuit(cirq.X(q1))
print(circuit)
q1.name = 'not_awesome'
print(circuit)
```

returns

```
dataclasses.FrozenInstanceError: cannot assign to field 'name'
```

Pylint & pytest pass.

### GridQubit 
```
import cirq
q1 = cirq.GridQubit(1, 1)
circuit = cirq.Circuit(cirq.X(q1))
print(circuit)
q1.row = 2
print(circuit)
```

returns

```
dataclasses.FrozenInstanceError: cannot assign to field 'row'
```

Pylint & pytest pass.

### LineQubit

```
import cirq
q1 = cirq.LineQubit(1)
circuit = cirq.Circuit(cirq.X(q1))
print(circuit)
q1.x = 2
print(circuit)
```

returns

```
dataclasses.FrozenInstanceError: cannot assign to field 'x'
```

Pylint pass.
Pytest fails 2 tests due to the different hashes comparing with  `LineQid`, since this class is still mutable (they are different objects):

```
AssertionError: Items in the same group produced different hashes. Example: hash(cirq.LineQubit(0)) is 3430018387555 but hash(cirq.LineQid(0, dimension=2)) is 8601707190144160275.
```

If we try to remediate this, trying to add  `@dataclass(frozen=True)` to `LineQid`, we have a problem with `dimension` and the base class `_BaseLineQid`. 

So my question is: **are the eq hash tests important (do they need to be the same object?) so should we refactor this class to handle `dataclasses`?**

Another question: **should we handle `dataclasses.FrozenInstanceError`, and if yes, what's the best way?**

### CODE REVIEW

File: cirq/devices/line_qubit.py [hunk 1]
```diff
@@ -184,6 +186,8 @@ class LineQubit(_BaseLineQid):
         cirq.LineQubit(1)
     """
 
+    x: int
```
Comment (mpharrigan) [inline cirq/devices/line_qubit.py]: I don't think this is right considering the inheritance here.

File: cirq/ops/named_qubit.py [hunk 2]
```diff
@@ -23,13 +26,10 @@ class NamedQubit(raw_types.Qid):
...
     """
-
-    def __init__(self, name: str) -> None:
-        self._name = name
-        self._comp_key = _pad_digits(name)
+    _name: str
 
     def _comparison_key(self):
-        return self._comp_key
+        return _pad_digits(self._name)
```
Comment (mpharrigan) [inline cirq/ops/named_qubit.py]: caching this value was likely done for performance reasons. Do you have a sense for the slowdown one would expect from this change?

File: cirq/devices/line_qubit.py (hunk 1 above)
Comment (cduck) [reply cirq/devices/line_qubit.py]: I think all three classes here should be dataclasses with `_BaseLineQid` defining the `x` field and `LineQid` defining the `dimension` field.

File: cirq/ops/named_qubit.py (hunk 2 above)
Comment (Strilanc) [reply cirq/ops/named_qubit.py]: Yes, this caching should not be removed.

**Comment by vtomole** [comment]
Ping @bt3gl

**Comment by bt3gl** [comment]
> Ping @bt3gl

Hi @vtomole I am sorry this PR got carried way. I will be working on the feedbacks and cleaning up everything this Saturday

**Comment by bt3gl** [comment]
Sorry, folks, I didn't have the chance to work on this in a timely manner so I will be just closing the PR.

**Pull request closed without merging**